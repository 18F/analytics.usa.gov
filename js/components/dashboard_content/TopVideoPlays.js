import React, { useRef, useEffect, useState } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import ChartBuilder from "../../lib/chart_helpers/chart_builder";
import DataLoader from "../../lib/data_loader";
import barChart from "../../lib/chart_helpers/barchart";
import { exceptions, titleExceptions } from "../../lib/exceptions";
import formatters from "../../lib/chart_helpers/formatters";

/**
 * Retrieves the top downloads report from the passed data URL and creates a
 * visualization for the count of users downloading files for the current
 * agency.
 *
 * @param {object} props the properties for the component
 * @param {string} props.dataHrefBase the URL of the base location of the data
 * to be downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 * @param {string} props.reportFileName the file name of the report to use as a
 * data source for the data visualization.
 * @param {number} props.numberOfListingsToDisplay the count of downloads
 * listings to display in the bar chart.
 */
function TopVideoPlays({
  dataHrefBase,
  reportFileName,
  numberOfListingsToDisplay,
}) {
  const dataURL = `${dataHrefBase}/${reportFileName}`;
  const ref = useRef(null);
  const [videoPlayData, setVideoPlayData] = useState(null);

  useEffect(() => {
    const initVideoPlaysChart = async () => {
      if (!videoPlayData) {
        const data = await DataLoader.loadJSON(dataURL);
        await setVideoPlayData(data);
      } else {
        const chartBuilder = new ChartBuilder();
        await chartBuilder
          .setElement(ref.current)
          .setData(videoPlayData)
          .setTransformer((d) => d.data.slice(0, numberOfListingsToDisplay))
          .setRenderer((selection) => {
            const barchartRenderer = barChart()
              .label((d) => d.video_title)
              .value((d) => +d.total_events)
              .scale((values) =>
                d3.scale
                  .linear()
                  .domain([0, 1, d3.max(values)])
                  .rangeRound([0, 1, 100]),
              )
              .format(formatters.addCommas);

            barchartRenderer(selection);

            // turn the labels into links
            selection
              .selectAll(".label")
              .each(function (d) {
                d.text = this.innerText;
              })
              .html("")
              .append("a")
              .attr("target", "_blank")
              .attr("rel", "noopener")
              .attr("href", (d) => exceptions[d.video_url] || d.video_url)
              .text((d) => titleExceptions[d.video_title] || d.video_title);
          })
          .build();
      }
    };
    initVideoPlaysChart().catch(console.error);
  }, [videoPlayData]);

  return (
    <figure className="top-video-plays__bar-chart" ref={ref}>
      <div className="data bar-chart"></div>
    </figure>
  );
}

TopVideoPlays.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
  reportFileName: PropTypes.string.isRequired,
  numberOfListingsToDisplay: PropTypes.number.isRequired,
};

export default TopVideoPlays;
