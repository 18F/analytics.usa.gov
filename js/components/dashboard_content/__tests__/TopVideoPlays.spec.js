import React from "react";
import { render, screen, waitFor } from "@testing-library/react";

import { delay } from "../../../../spec/support/test_utilities";
import DataLoader from "../../../lib/data_loader";
import TopVideoPlays from "../TopVideoPlays";

jest.mock("../../../lib/data_loader", () => ({
  ...jest.requireActual("../../../lib/data_loader"),
  loadDailyReportJSON: jest.fn(),
}));

describe("TopVideoPlays", () => {
  let component;
  let data;

  beforeEach(async () => {
    data = {
      name: "top-video-plays-30-days",
      query: {},
      meta: {
        name: "Top video plays 30 days",
        description: "Top 20 video plays in the last 30 days",
      },
      data: [
        {
          event_label: "video_start",
          video_title:
            "NASA Astronauts Send Fourth of July Wishes From the International Space Station",
          video_url: "https://www.youtube.com/watch?v=ijkkPoXQukY",
          total_events: "22112",
        },
        {
          event_label: "video_start",
          video_title: "Animation: The James Webb Space Telescope's Orbit",
          video_url: "https://www.youtube.com/watch?v=6cUe4oMk69E",
          total_events: "17997",
        },
        {
          event_label: "video_start",
          video_title: "How to Renew Your U.S.  Passport",
          video_url: "https://www.youtube.com/watch?v=0akd7xj6wec",
          total_events: "7568",
        },
        {
          event_label: "video_start",
          video_title:
            "Applying in Person for a U.S. Passport 2022 [Audio Description]",
          video_url: "https://www.youtube.com/watch?v=FsQHz2ZdX6E",
          total_events: "3485",
        },
        {
          event_label: "video_start",
          video_title: "HRSA's Ryan White HIV/AIDS Program 30th Anniversary",
          video_url: "https://www.youtube.com/watch?v=iQrP4n-VeFg",
          total_events: "1808",
        },
        {
          event_label: "video_start",
          video_title: "Primary Mirror Size Comparison Between Webb and Hubble",
          video_url: "https://www.youtube.com/watch?v=j3mk6tUokm4",
          total_events: "877",
        },
        {
          event_label: "video_start",
          video_title: "Online Passport Renewal: Uploading Your Photo",
          video_url: "https://www.youtube.com/watch?v=N33FJaziUgY",
          total_events: "859",
        },
        {
          event_label: "video_start",
          video_title: "Animation: The James Webb Space Telescope's Orbit",
          video_url:
            "https://www.youtube.com/watch?list=TLGG8tIphgpDAHkyNTA2MjAyNA&v=6cUe4oMk69E",
          total_events: "637",
        },
        {
          event_label: "video_start",
          video_title:
            "Animation of Asteroid Apophis’ 2029 Close Approach with Earth",
          video_url: "https://www.youtube.com/watch?v=hjJIyZKbHqc",
          total_events: "390",
        },
        {
          event_label: "video_start",
          video_title: "Animation: The James Webb Space Telescope's Orbit",
          video_url:
            "https://www.youtube.com/watch?list=TLGG8tIphgpDAHkyMzA2MjAyNA&v=6cUe4oMk69E",
          total_events: "343",
        },
        {
          event_label: "video_start",
          video_title: "Animation: The James Webb Space Telescope's Orbit",
          video_url:
            "https://www.youtube.com/watch?list=TLGG8tIphgpDAHkxOTA2MjAyNA&v=6cUe4oMk69E",
          total_events: "257",
        },
        {
          event_label: "video_start",
          video_title: "Animation: The James Webb Space Telescope's Orbit",
          video_url:
            "https://www.youtube.com/watch?list=TLGG8tIphgpDAHkxMzA2MjAyNA&v=6cUe4oMk69E",
          total_events: "224",
        },
        {
          event_label: "video_start",
          video_title: "Animation: The James Webb Space Telescope's Orbit",
          video_url:
            "https://www.youtube.com/watch?list=TLGG8tIphgpDAHkyMDA2MjAyNA&v=6cUe4oMk69E",
          total_events: "211",
        },
        {
          event_label: "video_start",
          video_title:
            "Pillars of Creation Star in New Visualization from NASA’s Hubble and Webb Telescopes",
          video_url: "https://www.youtube.com/watch?v=9ZooCy59rV0",
          total_events: "201",
        },
        {
          event_label: "video_start",
          video_title: "HRSA's Ryan White HIV/AIDS Program 30th Anniversary",
          video_url:
            "https://www.youtube.com/watch?list=TLGG8F2js94CWDcyNDA2MjAyNA&v=iQrP4n-VeFg",
          total_events: "166",
        },
        {
          event_label: "video_start",
          video_title: "Animation: The James Webb Space Telescope's Orbit",
          video_url:
            "https://www.youtube.com/watch?list=TLGG8tIphgpDAHkyNDA2MjAyNA&v=6cUe4oMk69E",
          total_events: "154",
        },
        {
          event_label: "video_start",
          video_title: "HRSA's Ryan White HIV/AIDS Program 30th Anniversary",
          video_url:
            "https://www.youtube.com/watch?list=TLGG8F2js94CWDcyMTA2MjAyNA&v=iQrP4n-VeFg",
          total_events: "151",
        },
        {
          event_label: "video_start",
          video_title: "Antarctic Ice Mass Loss 2002-2023",
          video_url: "https://www.youtube.com/watch?v=6PLBEz7o6wc",
          total_events: "146",
        },
        {
          event_label: "video_start",
          video_title: "Primary Mirror Size Comparison Between Webb and Hubble",
          video_url:
            "https://www.youtube.com/watch?list=TLGG9AelNnPkoFQxODA2MjAyNA&v=j3mk6tUokm4",
          total_events: "146",
        },
        {
          event_label: "video_start",
          video_title: "Global Warming from 1880 to 2022",
          video_url: "https://www.youtube.com/watch?v=LwRTw_7NNJs",
          total_events: "140",
        },
      ],
      totals: {},
      taken_at: "2024-07-11T14:36:05.272Z",
    };
    DataLoader.loadDailyReportJSON.mockImplementation(() => {
      return Promise.resolve(data);
    });

    component = render(
      <TopVideoPlays
        agency={"Interior"}
        dataHrefBase={"http://www.example.com/data/"}
        numberOfListingsToDisplay={10}
      />,
    );
    await waitFor(() =>
      screen.getByRole("link", {
        name: "NASA Astronauts Send Fourth of July Wishes From the International Space Station",
      }),
    );
    // Wait for barchart transition animation to complete (200 ms, set in
    // js/lib/chart_helpers/barchart.js)
    await delay(300);
  });

  it("renders a component with data loaded", () => {
    expect(component.asFragment()).toMatchSnapshot();
  });
});
