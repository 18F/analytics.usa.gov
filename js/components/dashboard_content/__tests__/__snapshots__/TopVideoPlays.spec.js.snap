// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopVideoPlays when data is loaded renders a component with data loaded 1`] = `
<DocumentFragment>
  <figure
    class="top-video-plays__bar-chart"
  >
    <div
      class="data bar-chart"
    >
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <a
            href="https://www.youtube.com/watch?v=ijkkPoXQukY"
            rel="noopener"
            target="_blank"
          >
            NASA Astronauts Send Fourth of July Wishes From the International Space Station
          </a>
        </div>
        <div
          class="value"
        >
          22,112
        </div>
        <div
          class="bar"
          style="width: 100.0%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <a
            href="https://www.youtube.com/watch?v=6cUe4oMk69E"
            rel="noopener"
            target="_blank"
          >
            Animation: The James Webb Space Telescope's Orbit
          </a>
        </div>
        <div
          class="value"
        >
          17,997
        </div>
        <div
          class="bar"
          style="width: 82.0%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <a
            href="https://www.youtube.com/watch?v=0akd7xj6wec"
            rel="noopener"
            target="_blank"
          >
            How to Renew Your U.S.  Passport
          </a>
        </div>
        <div
          class="value"
        >
          7,568
        </div>
        <div
          class="bar"
          style="width: 35.0%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <a
            href="https://www.youtube.com/watch?v=FsQHz2ZdX6E"
            rel="noopener"
            target="_blank"
          >
            Applying in Person for a U.S. Passport 2022 [Audio Description]
          </a>
        </div>
        <div
          class="value"
        >
          3,485
        </div>
        <div
          class="bar"
          style="width: 17.0%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <a
            href="https://www.youtube.com/watch?v=iQrP4n-VeFg"
            rel="noopener"
            target="_blank"
          >
            HRSA's Ryan White HIV/AIDS Program 30th Anniversary
          </a>
        </div>
        <div
          class="value"
        >
          1,808
        </div>
        <div
          class="bar"
          style="width: 9.0%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <a
            href="https://www.youtube.com/watch?v=j3mk6tUokm4"
            rel="noopener"
            target="_blank"
          >
            Primary Mirror Size Comparison Between Webb and Hubble
          </a>
        </div>
        <div
          class="value"
        >
          877
        </div>
        <div
          class="bar"
          style="width: 5.0%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <a
            href="https://www.youtube.com/watch?v=N33FJaziUgY"
            rel="noopener"
            target="_blank"
          >
            Online Passport Renewal: Uploading Your Photo
          </a>
        </div>
        <div
          class="value"
        >
          859
        </div>
        <div
          class="bar"
          style="width: 5.0%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <a
            href="https://www.youtube.com/watch?list=TLGG8tIphgpDAHkyNTA2MjAyNA&v=6cUe4oMk69E"
            rel="noopener"
            target="_blank"
          >
            Animation: The James Webb Space Telescope's Orbit
          </a>
        </div>
        <div
          class="value"
        >
          637
        </div>
        <div
          class="bar"
          style="width: 4.0%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <a
            href="https://www.youtube.com/watch?v=hjJIyZKbHqc"
            rel="noopener"
            target="_blank"
          >
            Animation of Asteroid Apophisâ€™ 2029 Close Approach with Earth
          </a>
        </div>
        <div
          class="value"
        >
          390
        </div>
        <div
          class="bar"
          style="width: 3.0%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <a
            href="https://www.youtube.com/watch?list=TLGG8tIphgpDAHkyMzA2MjAyNA&v=6cUe4oMk69E"
            rel="noopener"
            target="_blank"
          >
            Animation: The James Webb Space Telescope's Orbit
          </a>
        </div>
        <div
          class="value"
        >
          343
        </div>
        <div
          class="bar"
          style="width: 3.0%;"
        />
      </div>
    </div>
  </figure>
</DocumentFragment>
`;

exports[`TopVideoPlays when data is not loaded renders a component in loading state 1`] = `
<DocumentFragment>
  <figure
    class="top-video-plays__bar-chart"
  >
    <div
      class="data bar-chart"
    />
  </figure>
</DocumentFragment>
`;

exports[`TopVideoPlays when data loading has an error renders a component in error state 1`] = `
<DocumentFragment>
  <figure
    class="top-video-plays__bar-chart"
  >
    <div
      class="data bar-chart"
    />
  </figure>
</DocumentFragment>
`;
